runtime/runtime/src/prefetch.rs

```diff
@@ -142,10 +142,17 @@ impl TriePrefetcher {
                 }
             }
 
+            let mut code_prefetch_requested = false;
             for action in &action_receipt.actions {
                 let Action::FunctionCall(fn_call) = action else {
                     continue;
                 };
+                if !code_prefetch_requested {
+                    let trie_key = TrieKey::ContractCode { account_id: account_id.clone() };
+                    self.prefetch_trie_key(trie_key)?;
+                    code_prefetch_requested = true;
```

Comment:
If we're prefetching the contract code, we might as well remove the optimization that was introduced to not read the code when not necessary, yeah?

---

What happens if the prefetch request gets cancelled (i.e. is not finished in time)? This also needs some measurement on how it might affect the throughput/performance of the receipts that currently have custom prefetch rules.

Reply:
I don't think we should remove it. As Bowen mentioned on some meeting, if we ever need to read code from storage, then we need to compile it, and it will significantly slow down chunk execution. So it looks like we need to precompile all contracts before chunk execution and avoid any get_code calls.

For stateless validation, we have an idea to distribute contracts separately from state witness as they are taking huge part of storage proofs anyway.

If prefetch doesn't finish in time, we just read code on `Trie::update`, if we record storage. For MVP the forknet performance looked fine, but later we need to improve that.

Reply:
> if we ever need to read code from storage, then we need to compile it

This implication is reversed -- we only ever need to read the code from storage if there's a need to compile something. But we aren't required to also go into the compilation logic if the code happens to be read out.

> If prefetch doesn't finish in time, we just read code on Trie::update, if we record storage. For MVP the forknet performance looked fine, but later we need to improve that.

I'm a little concerned about how it would affect the other prefetch requests for the same receipt. We've started relying on the prefetcher quite a bit at this point, and if fetching the code prevents the contract data from being prefetched in time, then it may very well have a negative effect on the performance of the node especially for shard 2 wrt sweat contracts. But if forknet worked fine...

Reply:
> And my suggestion was that we could go back to doing so if we can figure out a way to make it not block the rest of the flow, and supposedly prefetching the contract code would do that to an extent...

Yeah, the idea is to move this work to prefetcher. And also, there is a future plan for chunk validators to request contract codes from chunk producers separately and precompile them, so Runtime shouldn't be aware of codes.
Another reason is - we can't use regular flow of reading code from trie anymore, because if it is non-blocking, then codes will be recorded in state witness at random times. And we need to build state witness deterministically because we rely on its current size to decide whether to execute new receipt or not / fail function call or not.

Forknet worked fine. Actually we use in-memory trie there, so we can even drop prefetching of contract data there, leaving only prefetching of large values (contract codes)